src/parser/monk/mistweaver/modules/talents/InvokeChiJi.js

```diff
@@ -1,33 +1,136 @@
+  efChannelStart = 0;
+  efChannelEnd = 0;
+  
+  envCount = 0;
+  
   constructor(...args) {
     super(...args);
     this.active = this.selectedCombatant.hasTalent(SPELLS.INVOKE_CHIJI_THE_RED_CRANE_TALENT.id);
     if (!this.active) {return;}
     this.addEventListener(Events.heal.by(SELECTED_PLAYER).spell(SPELLS.GUST_OF_MISTS_CHIJI), this.handleGust);
     this.addEventListener(Events.heal.by(SELECTED_PLAYER).spell(SPELLS.ENVELOPING_BREATH), this.handleEnvelopingBreath);
+    this.addEventListener(Events.cast.by(SELECTED_PLAYER).spell(SPELLS.ENVELOPING_MIST), this.handleEnvelopCast);
+    this.addEventListener(Events.cast.by(SELECTED_PLAYER).spell(SPELLS.INVOKE_CHIJI_THE_RED_CRANE_TALENT), this.handleChijiStart);
+    this.addEventListener(Events.death.to(SELECTED_PLAYER_PET), this.handleChijiDeath);
+    this.addEventListener(Events.death.to(SELECTED_PLAYER), this.handleChijiDeath);
+    this.addEventListener(Events.damage.by(SELECTED_PLAYER).spell([SPELLS.BLACKOUT_KICK, SPELLS.RISING_SUN_KICK_SECOND, SPELLS.BLACKOUT_KICK_TOTM]), this.handleOvercapStacks)
+    //need a different eventlistener beacause chiji currently only applies 1 stack per cast of sck, not on each dmg event
+    this.addEventListener(Events.cast.by(SELECTED_PLAYER).spell(SPELLS.SPINNING_CRANE_KICK), this.handleOvercapStacks);
+    this.addEventListener(Events.GlobalCooldown.by(SELECTED_PLAYER), this.handleGlobal);
+    this.addEventListener(Events.BeginChannel.by(SELECTED_PLAYER).spell(SPELLS.ESSENCE_FONT), this.handleEssenceFontStart);
+    this.addEventListener(Events.EndChannel.by(SELECTED_PLAYER).spell(SPELLS.ESSENCE_FONT), this.handleEssenceFontEnd);
+  }
+
+  //missed gcd mangement
+  handleChijiStart(event) {
+    this.chijiActive = true;
+    this.chijiStart = this.lastGlobal = event.timestamp;
+    this.chijiGlobals += 1;
+    this.chijiUses +=1;
+  }
+  
+  handleChijiDeath(event) {
+    this.chijiActive = false;
+  }
+
+  handleGlobal(event){
+    if(this.chijiActive) {
+      this.chijiGlobals += 1;
+      //if timebetween globals is longer than the gcd add the difference to the missed gcd tally
+      //we only care about accounting for channels of essence font during WoO, other than that it should be the gcd during chiji
+      if(event.ability.guid === SPELLS.ESSENCE_FONT.id && this.selectedCombatant.hasBuff(SPELLS.WEAPONS_OF_ORDER_BUFF_AND_HEAL.id)) {
+        return;
+      } else if((event.timestamp - this.lastGlobal) > event.duration) {
+        this.missedGlobals += (event.timestamp - this.lastGlobal - event.duration)/event.duration;
+      }
+      this.lastGlobal = event.timestamp;
+    }
+  }
+
+  handleEssenceFontStart(event) { 
+    if(this.chijiActive && this.selectedCombatant.hasBuff(SPELLS.WEAPONS_OF_ORDER_BUFF_AND_HEAL.id)) { 
+       this.efChannelStart = this.lastGlobal = event.timestamp;
+    }
+  }
+
+  handleEssenceFontEnd(event) {
+    const gcd = 1500 / (1 + this.statTracker.hastePercentage(this.statTracker.currentHasteRating));
```

Comment:
iirc ef's channel time is snapshotted at the haste you have at the start of the channel as well as the gcd starting at the start of the channel
On top of this iirc essence font itself has a GCD event where you can just grab the gcd from
So I might have miss lead you with start and end channel since you can do it with
gcd and end channel

Reply:
updated to use cast and end channel